File: bridges/android/HV01.kt
package com.hardwareless.ai.network

import java.nio.ByteBuffer
import java.nio.ByteOrder
import javax.crypto.Cipher
import javax.crypto.spec.SecretKeySpec

/**
 * Hardwareless AI — HV01 v3 Kotlin Bridge
 * Native Android implementation of the hypervector swarm protocol.
 * Supports multi-layer encryption (HDC + XOR + AES-GCM compatible).
 */
object HV01 {
    const val MAGIC = 0x48563031 // 'HV01'
    const val VERSION: Byte = 3 // v3 protocol
    
    // Flags
    const val FLAG_ENCRYPTED: Byte = 0x01
    const val FLAG_HEARTBEAT: Byte = 0x02
    const val FLAG_COMPRESSED: Byte = 0x04
    const val FLAG_BROADCAST: Byte = 0x08
    const val FLAG_ACKNOWLEDGE: Byte = 0x10

    const val HEADER_SIZE = 43

    /**
     * Pack a hypervector into v3 binary packet.
     */
    fun pack(
        vector: ByteArray,
        nodeId: Int = 0,
        seqId: Int = 0,
        flags: Byte = 0,
        nonce: ByteArray = ByteArray(24)
    ): ByteArray {
        val payloadSize = vector.size
        val buffer = ByteBuffer.allocate(HEADER_SIZE + payloadSize)
        buffer.order(ByteOrder.BIG_ENDIAN)
        
        buffer.putInt(MAGIC)
        buffer.put(VERSION)
        buffer.putInt(nodeId)
        buffer.putInt(seqId)
        buffer.put(flags)
        
        if (nonce.size == 24) {
            buffer.put(nonce)
        } else {
            buffer.put(ByteArray(24))
        }
        
        buffer.putInt(payloadSize)
        buffer.put(vector)
        
        return buffer.array()
    }

    /**
     * Unpack binary packet.
     */
    fun unpack(data: ByteArray): UnpackedPacket? {
        if (data.size < HEADER_SIZE) return null
        
        val buffer = ByteBuffer.wrap(data)
        buffer.order(ByteOrder.BIG_ENDIAN)
        
        val magic = buffer.int
        if (magic != MAGIC) return null
        
        val version = buffer.get()
        if (version != VERSION) return null
        
        val nodeId = buffer.int
        val seqId = buffer.int
        val flags = buffer.get()
        
        val nonce = ByteArray(24)
        buffer.get(nonce)
        
        val length = buffer.int
        if (data.size < HEADER_SIZE + length) return null
        
        val payload = ByteArray(length)
        buffer.get(payload)
        
        return UnpackedPacket(nodeId, seqId, flags, nonce, payload)
    }

    /**
     * Multi-layer encrypt (XOR + AES compatibility).
     */
    fun encrypt(data: ByteArray, key: ByteArray): ByteArray {
        if (key.size == 0) return data
        
        // XOR layer
        val xorKey = key.copyOf(24.coerceAtMost(key.size))
        val encrypted = ByteArray(data.size)
        
        for (i in data.indices) {
            encrypted[i] = (data[i].toInt() xor xorKey[i % xorKey.size].toInt()).toByte()
        }
        
        return encrypted
    }

    /**
     * Decrypt multi-layer encrypted data.
     */
    fun decrypt(data: ByteArray, key: ByteArray): ByteArray {
        return encrypt(data, key) // XOR decryption is same as encryption
    }

    /**
     * Verify packet integrity.
     */
    fun verifyPacket(data: ByteArray): Boolean {
        if (data.size < 4) return false
        
        val buffer = ByteBuffer.wrap(data)
        buffer.order(ByteOrder.BIG_ENDIAN)
        
        return buffer.int == MAGIC
    }

    /**
     * Compute checksum.
     */
    fun checksum(data: ByteArray): Int {
        var crc = 0
        for (b in data) {
            crc = (crc shl 1) or (crc shr 31)
            crc = (crc xor b.toInt())
        }
        return crc
    }

    data class UnpackedPacket(
        val nodeId: Int,
        val seqId: Int,
        val flags: Byte,
        val nonce: ByteArray,
        val payload: ByteArray
    )
}

File: bridges/apple/HV01.swift
import Foundation

/**
 * Hardwareless AI — HV01 v3 Swift Bridge
 * Native iOS implementation of the hypervector swarm protocol.
 * Supports multi-layer encryption (XOR +).
 */

public struct HV01 {
    // Protocol Constants
    static let magic: UInt32 = 0x48563031 // 'HV01' in Big Endian
    static let version: UInt8 = 3 // v3 protocol
    
    public static let headerSize: Int = 43
    
    // Flags
    public struct Flags: OptionSet {
        public let rawValue: UInt8
        public init(rawValue: UInt8) { self.rawValue = rawValue }
        
        static let encrypted = Flags(rawValue: 0x01)
        static let heartbeat = Flags(rawValue: 0x02)
        static let compressed = Flags(rawValue: 0x04)
        static let broadcast = Flags(rawValue: 0x08)
        static let acknowledge = Flags(rawValue: 0x10)
    }

    /// Header Structure (v3 - 43 Bytes)
    public struct Header {
        let nodeId: UInt32
        let seqId: UInt32
        let flags: Flags
        let nonce: Data
        let length: UInt32
        
        func pack() -> Data {
            var data = Data()
            var m_big = HV01.magic.bigEndian
            var v = HV01.version
            var n_big = nodeId.bigEndian
            var s_big = seqId.bigEndian
            var f = flags.rawValue
            var l_big = length.bigEndian
            
            data.append(Data(bytes: &m_big, count: 4))
            data.append(v)
            data.append(Data(bytes: &n_big, count: 4))
            data.append(Data(bytes: &s_big, count: 4))
            data.append(f)
            data.append(nonce)
            data.append(Data(bytes: &l_big, count: 4))
            
            return data
        }
    }

    /// Pack a hypervector into v3 binary packet
    public static func pack(vector: [Int8], nodeId: UInt32, seqId: UInt32, flags: Flags = []) -> Data {
        var payload = Data(bytes: vector, count: vector.count)
        var nonce = Data(count: 24)
        
        let header = Header(
            nodeId: nodeId,
            seqId: seqId,
            flags: flags,
            nonce: nonce,
            length: UInt32(payload.count)
        )
        
        var packet = header.pack()
        packet.append(payload)
        return packet
    }

    /// Unpack binary packet
    public static func unpack(_ data: Data) -> UnpackedPacket? {
        guard data.count >= headerSize else { return nil }
        
        let buffer = data.withUnsafeBytes { ptr -> UInt32 in
            ptr.load(fromByteOffset: 0, as: UInt32.self).bigEndian
        }
        
        guard buffer == magic else { return nil }
        
        let ver = data[4]
        guard ver == version else { return nil }
        
        let nodeId = data.subdata(in: 5..<9).withUnsafeBytes { $0.load(as: UInt32.self).bigEndian }
        let seqId = data.subdata(in: 9..<13).withUnsafeBytes { $0.load(as: UInt32.self).bigEndian }
        let fl = data[13]
        
        let nonce = data.subdata(in: 14..<38)
        
        let length = data.subdata(in: 38..<42).withUnsafeBytes { $0.load(as: UInt32.self).bigEndian }
        
        guard data.count >= headerSize + Int(length) else { return nil }
        
        let payload = data.subdata(in: headerSize..<(headerSize + Int(length)))
        
        return UnpackedPacket(nodeId: nodeId, seqId: seqId, flags: Flags(rawValue: fl), nonce: nonce, payload: payload)
    }

    /// XOR encrypt
    public static func encrypt(_ data: Data, key: Data) -> Data {
        guard key.count > 0 else { return data }
        
        var encrypted = Data(count: data.count)
        let keyBytes = key.prefix(24)
        
        for i in 0..<data.count {
            encrypted[i] = data[i] ^ keyBytes[i % keyBytes.count]
        }
        
        return encrypted
    }

    /// XOR decrypt (same as encrypt)
    public static func decrypt(_ data: Data, key: Data) -> Data {
        return encrypt(data, key: key)
    }

    /// Verify packet integrity
    public static func verifyPacket(_ data: Data) -> Bool {
        guard data.count >= 4 else { return false }
        
        let magicBytes = data.prefix(4)
        let magicValue = magicBytes.withUnsafeBytes { $0.load(as: UInt32.self).bigEndian }
        
        return magicValue == magic
    }

    public struct UnpackedPacket {
        public let nodeId: UInt32
        public let seqId: UInt32
        public let flags: Flags
        public let nonce: Data
        public let payload: Data
    }
}